### PR TITLE
bpf, docs: Add initial BPF ELF profile draft

### PR DESCRIPTION
**Pull request by dthaler**
Add initial BPF ELF profile draft

Signed-off-by: Dave Thaler <dthaler@microsoft.com>

### CODE REVIEW

**Comment by dthaler** [comment]
@qmonnet Can you review?  Also, this draft uses IETF RFC format for structs rather than the style used in instruction-set.rst, any comments on whether this is a better or worse direction?

**Comment by qmonnet** [comment]
I'll look at it next week, quickly browsing through the doc I noticed that some of the conventions do not reflect the changes acted in libbpf v1.0 (see dropping support for `maps` section in favour of `.maps` ([Andrii's blog](https://nakryiko.com/posts/libbpf-v1/#other-api-clean-ups-renames-removals-etc)), switching to fixed section names for program types instead of longest prefix match ([Andrii's blog again](https://nakryiko.com/posts/libbpf-v1/#bpf-program-sec-annotation-streamlining))). Not sure what should be the "official" convention, but libbpf v1 should likely remain compliant.

**Comment by dthaler** [comment]
> I'll look at it next week, quickly browsing through the doc I noticed that some of the conventions do not reflect the changes acted in libbpf v1.0 (see dropping support for `maps` section in favour of `.maps` ([Andrii's blog](https://nakryiko.com/posts/libbpf-v1/#other-api-clean-ups-renames-removals-etc)), switching to fixed section names for program types instead of longest prefix match ([Andrii's blog again](https://nakryiko.com/posts/libbpf-v1/#bpf-program-sec-annotation-streamlining))). Not sure what should be the "official" convention, but libbpf v1 should likely remain compliant.

Thanks for the pointers!   Is there a spec for or description of the ".maps" section somewhere?  Or just the code in libbpf.c?

**Comment by qmonnet** [comment]
> Thanks for the pointers! Is there a spec for or description of the ".maps" section somewhere?

Not that I know of :/.

> Or just the code in libbpf.c?

Sounds likely, unless some of the other libraries (cilium/ebpf, Aya) has more docs, but I haven't checked so far. Also not sure what's their status regarding section names conventions or `maps` vs. `.maps` for example.

**Comment by dthaler** [comment]
> > Thanks for the pointers! Is there a spec for or description of the ".maps" section somewhere?
> 
> Not that I know of :/.
> 
> > Or just the code in libbpf.c?
> 
> Sounds likely, unless some of the other libraries (cilium/ebpf, Aya) has more docs, but I haven't checked so far. Also not sure what's their status regarding section names conventions or `maps` vs. `.maps` for example.

cilium/ebpf seems to read _both_ types (and also data sections), which seems reasonable behavior:
https://github.com/cilium/ebpf/blob/master/elf_reader.go#L125

Aya also seems to read both types:
https://github.com/aya-rs/aya/blob/main/aya/src/obj/mod.rs#L1024

redbpf (which unlike Aya is listed as Major) only supports "maps":
https://github.com/foniod/redbpf/blob/main/redbpf/src/lib.rs#L1544

File: rst/linux-notes.rst [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+Arithmetic instructions
+--------------------------------
+
+While the eBPF instruction set document uses the standard C terminology as the cross-platform specification,
+in the Linux kernel, uint32_t is expressed as u32, uint64_t is expressed as u64, etc.
+
+Byte swap instructions
+--------------------------------
+
+``BPF_FROM_LE`` and ``BPF_FROM_BE`` exist as aliases for ``BPF_TO_LE`` and ``BPF_TO_BE`` respectively.
```
Comment (matt-gretton-dann) [inline rst/linux-notes.rst]: Is this the right way round?
```suggestion
``BPF_FROM_LE`` and ``BPF_FROM_BE`` exist as aliases for ``BPF_TO_BE`` and ``BPF_TO_LE`` respectively.
```

File: rst/instruction-set.rst [hunk 2]
```diff
@@ -11,198 +23,328 @@ all of which are 64-bits wide.
...
-
-The eBPF program needs to store the return value into register R0 before doing a
-BPF_EXIT.
 
+The 4-bit 'code' field encodes the operation as below, where PC is the program counter:
+
+========  =====  ===  ==========================  ========================
+code      value  src  description                 notes
+========  =====  ===  ==========================  ========================
+BPF_JA    0x0    0x0  PC += offset                BPF_JMP only
```
Comment (matt-gretton-dann) [inline rst/instruction-set.rst]:
```suggestion
BPF_JA    0x0    0x0  PC += offset                BPF_JMP, BPF_JMP32 only
```

Add `BPF_JMP32` as well?

File: rst/instruction-set.rst [hunk 3]
```diff
@@ -11,198 +23,328 @@ all of which are 64-bits wide.
...
+BPF_JGT   0x2    any  PC += offset if dst > src   unsigned
+BPF_JGE   0x3    any  PC += offset if dst >= src  unsigned
+BPF_JSET  0x4    any  PC += offset if dst & src
+BPF_JNE   0x5    any  PC += offset if dst != src
+BPF_JSGT  0x6    any  PC += offset if dst > src   signed
+BPF_JSGE  0x7    any  PC += offset if dst >= src  signed
+BPF_CALL  0x8    0x0  call helper function imm    see `Helper functions`_
+BPF_CALL  0x8    0x1  call PC += offset           see `eBPF functions`_
+BPF_CALL  0x8    0x2  call runtime function imm   see `Runtime functions`_
+BPF_EXIT  0x9    0x0  return                      BPF_JMP only
```
Comment (matt-gretton-dann) [inline rst/instruction-set.rst]:
```suggestion
BPF_EXIT  0x9    0x0  return                      BPF_JMP, BPF_JMP32 only
```

File: rst/elf.rst [hunk 4]
```diff
@@ -0,0 +1,409 @@
+.. contents::
+.. sectnum::
+
+====================================
+eBPF ELF Profile Specification, v0.1
+====================================
+
+The Executable and Linking Format (ELF) is specified in
+Tool Interface Standard (TIS), "Executable and Linking Format (ELF) Specification, Version 1.2", May 1995, https://refspecs.linuxbase.org/elf/elf.pdf.
```
Comment (matt-gretton-dann) [inline rst/elf.rst]: This is an older document: maybe pick either the latest (2012) or draft version (2013) from here: http://www.sco.com/developers/gabi/

In particular the draft you refer to seems to be 32-bit only.

File: rst/elf.rst [hunk 5]
```diff
@@ -0,0 +1,409 @@
...
+
+This document specifies version 0.1 of the eBPF profile for ELF files.
+
+Documentation conventions
+=========================
+
+All integer fields are unsigned.
+
+TEXT Sections
```
Comment (matt-gretton-dann) [inline rst/elf.rst:33]: Does this document need to talk about what it expects in the ELF header, or how the ELF header affects the .text sections?  For instance do we need the machine field to be a magic `EM_BPF` value (do we have one assigned)?  Or do we expect BPF programs to also exist in the same file as other 'native' programs?

How do we expect 32-/64-bit to change .text sections?  Even if they don't let us be obvious about it.

Similarly big/little endian.  How does that affect instruction format?  Data format?

**Review by matt-gretton-dann — COMMENTED** [review]
A quick light-weight review.

File: rst/instruction-set.rst (hunk 2 above)
Comment (dthaler) [reply rst/instruction-set.rst]: According to https://elixir.bootlin.com/linux/latest/source/Documentation/bpf/instruction-set.rst#L173 and everything else I've seen, BPF_JMP32 is not permitted with BPF_JA.

File: rst/instruction-set.rst (hunk 3 above)
Comment (dthaler) [reply rst/instruction-set.rst]: According to https://elixir.bootlin.com/linux/latest/source/Documentation/bpf/instruction-set.rst#L182 and everything else I've seen, BPF_JMP32 is not permitted with BPF_EXIT.

File: rst/linux-notes.rst (hunk 1 above)
Comment (dthaler) [reply rst/linux-notes.rst]: You are correct, will fix.

File: rst/elf.rst (hunk 4 above)
Comment (dthaler) [reply rst/elf.rst]: Thanks for the pointer, will update.  FYI, https://en.wikipedia.org/wiki/Executable_and_Linkable_Format#References does not mention that link.

File: rst/elf.rst (hunk 5 above)
Comment (dthaler) [reply rst/elf.rst:33]: > Does this document need to talk about what it expects in the ELF header, or how the ELF header affects the .text sections? For instance do we need the machine field to be a magic `EM_BPF` value (do we have one assigned)?

Good point!  Yes one is defined and I have added a note about it.

> How do we expect 32-/64-bit to change .text sections? Even if they don't let us be obvious about it.

Only `ELFCLASS64` is permitted.  Added a note about it.

> Similarly big/little endian. How does that affect instruction format? Data format?

The .BTF and .BTF.ext sections already cover endianness detection if either of those sections are present.

File: rst/elf.rst (hunk 4 above)
Comment (matt-gretton-dann) [reply rst/elf.rst]: Indeed.  I got this form Arm's ABI docs https://github.com/ARM-software/abi-aa/blob/60a8eb8c55e999d74dac5e368fc9d7e36e38dda4/aaelf32/aaelf32.rst.

**Comment by dthaler** [comment]
Replacing this PR with PR #95

**Pull request closed without merging**